### PR TITLE
Stratified sampling

### PR DESCRIPTION
**Pull request by alishibli97**

### CODE REVIEW

File: pangaea/run.py [hunk 1]
```diff
@@ -147,13 +154,23 @@ def main(cfg: DictConfig) -> None:
                 preprocess, dataset=val_dataset, encoder=encoder
             )
         if 0 < cfg.limited_label < 1:
-            n_train_samples = len(train_dataset)
-            indices = random.sample(
-                range(n_train_samples), int(n_train_samples * cfg.limited_label)
-            )
-            train_dataset = Subset(train_dataset, indices)
+            # n_train_samples = len(train_dataset)
```
Comment (KerekesDavid) [inline pangaea/run.py:157]: You can remove these, that's why we have git :)

**Review by KerekesDavid — APPROVED** [review]
Good job, seems to be working with MADOS on my end.

One nitpick, since the stratification itself takes quite long, and one of the usecases for limiting dataset size is to have a quick test run on a few samples, could you add a toggle to have the old behavior of just random sampling?

**Comment by VMarsocci** [comment]
Hi before merging, please consider the following changes (already discussed in private- but I report them here for everyone else).
1. add the possibility to choose the stratified sampling on a single set (either train or val)
2. check the last version of HLSBurnScars (in the one you want to merge it is not updated)
3. replace the "calculate class distribution" function with just reading this distribution in the dataset config
Thanks a lot :)

**Comment by yurujaja** [comment]
To address the comments, some modifications are made: 
1. configure sampling for both train and val
2. configure the limited label strategy: random or stratified
3. the unused function in hlsburn scar is removed
4. the "calculate class distribution" function is still needed to calculate image-wise distribution to enable the selection of images

**Comment by alishibli97** [comment]
Added Regression stratification.
In train.yaml, one can specify the stratification method using the param:
`limited_label_strategy: stratified_classification  # or stratified_regression, or random`

@RituYadav92 is it possible to validate that it works on your side with biomastters?

**Comment by RituYadav92** [comment]
I made two modifications to the code:
1) Updated the code to select a fraction of labels from each bin. 
2) Updated variable names "labeled_idx, unlabeled_idx" to "selected_idx, other_idx"

Please update the same for classification if it fits well.

File: pangaea/utils/subset_sampler.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    # Step 3: Prep a dictionary to hold indices for each bin combination
+    indices_per_bin = {}
+
+    # Combine the bins for each class to create unique bin identifiers
+    combined_bins = np.apply_along_axis(lambda row: ''.join(map(str, row)), axis=1, arr=binned_distributions)
+
+    # Step 4: Populate the dictionary with indices based on combined bin identifiers
+    for idx, bin_id in enumerate(combined_bins):
+        if bin_id not in indices_per_bin:
```
Comment (RituYadav92) [inline pangaea/utils/subset_sampler.py:68]: Are you sure about "not in" ??

**Review by RituYadav92 — COMMENTED** [review]
The code looks fine now except line 68 in subset_sampler.py. It should be "if bin_id in indices_per_bin:" instead of "if bin_id not in indices_per_bin:"

Please check.

**Comment by alishibli97** [comment]
I think it is fine, what do you suspect? 
It is only to initialize the dict key if not present

**Comment by RituYadav92** [comment]
I see now, you didn't adapt the initialization from regression but did it other way. Np problem. Resolved.

**Review by VMarsocci — APPROVED** [review]
